### PR TITLE
Refactor Iterative Prompting Evaluation

### PR DESCRIPTION
**Pull request by anwai98**
WIP - adapting iterative prompting evaluation according to the updated iterative prompt generator.

### CODE REVIEW

File: micro_sam/evaluation/inference.py [hunk 1]
```diff
@@ -395,44 +394,47 @@ def _save_segmentation(masks, prediction_path):
...
+    multimasking = (n_pos == 1)
     prompt_generator = IterativePromptGenerator()
 
     n_samples = len(sampled_ids[0])
     n_batches = int(np.ceil(float(n_samples) / batch_size))
 
     for iteration in range(n_iterations):
-        final_masks = []
+        final_masks, all_sampled_binary_y = [], []
         for batch_idx in range(n_batches):
```
Comment (constantinpape) [inline micro_sam/evaluation/inference.py:432]: You should use https://github.com/computational-cell-analytics/micro-sam/blob/dev/micro_sam/inference.py#L17 here.

**Comment by anwai98** [comment]
This is working as expected (tested it for a few samples per setting). Let me know how it looks. Thanks!

File: micro_sam/evaluation/inference.py [hunk 2]
```diff
@@ -392,118 +389,127 @@ def _save_segmentation(masks, prediction_path):
...
     masks = [{"segmentation": mask, "area": mask.sum()} for mask in masks]
     segmentation = mask_data_to_segmentation(masks, shape, with_background=True)
-    imageio.imwrite(prediction_path, segmentation)
+    imageio.imwrite(prediction_path, segmentation, compression=5)
+
+
+def extract_instances_from_batched_outputs(batched_outputs):
+    masks = [m["segmentation"][None] for m in batched_outputs]
+    masks = torch.stack(masks)
+    masks = (masks > 0.5).to(torch.float32)
```
Comment (constantinpape) [inline micro_sam/evaluation/inference.py:398]: The masks are already binary, so you don't need this. Overall, this is simple enough to write it as a one-liner and then you don't need to refactor this into a function.

Comment (constantinpape) [reply micro_sam/evaluation/inference.py:398]: See comment below.

File: micro_sam/evaluation/inference.py [hunk 3]
```diff
@@ -392,118 +389,127 @@ def _save_segmentation(masks, prediction_path):
...
-                mask, l_mask = torch.stack(mask), torch.stack(l_mask)
-                masks.append(mask)
-                logits_masks.append(l_mask)
+        batched_outputs = batched_inference(
+            predictor, image, batch_size,
+            boxes=boxes, points=points, point_labels=point_labels,
+            multimasking=multimasking, embedding_path=embedding_path,
+            return_instance_segmentation=False
+        )
+        masks = extract_instances_from_batched_outputs(batched_outputs)
```
Comment (constantinpape) [inline micro_sam/evaluation/inference.py:450]: You can do this with a one-liner here:
```python
masks = torch.stack([m["segmentation"][None] for m in batched_outputs]).to(torch.float32)
```

File: micro_sam/evaluation/inference.py [hunk 4]
```diff
@@ -392,118 +389,127 @@ def _save_segmentation(masks, prediction_path):
...
+
+
+def extract_instances_from_batched_outputs(batched_outputs):
+    masks = [m["segmentation"][None] for m in batched_outputs]
+    masks = torch.stack(masks)
+    masks = (masks > 0.5).to(torch.float32)
+    return masks
+
+
+def get_channelwise_gt_objects(gt, gt_ids, device=None):
```
Comment (constantinpape) [inline micro_sam/evaluation/inference.py:402]: We have a function for this already and don't need this.

Comment (constantinpape) [reply micro_sam/evaluation/inference.py:402]: See comment below.

File: micro_sam/evaluation/inference.py [hunk 5]
```diff
@@ -392,118 +389,127 @@ def _save_segmentation(masks, prediction_path):
...
+        gt, gt_ids, use_points, use_boxes,
+        n_positives=n_positives,
+        n_negatives=n_negatives,
+        dilation=dilation
     )
-    gt = torch.from_numpy(gt[None].astype("int32"))
 
-    n_pos = 0 if use_boxes else 1
-    batched_inputs, sampled_ids = to_sam_inputs(image, gt, n_pos=n_pos, n_neg=0, get_boxes=use_boxes)
+    sampled_binary_gt = get_channelwise_gt_objects(gt, gt_ids)
```
Comment (constantinpape) [inline micro_sam/evaluation/inference.py:436]: You can use https://github.com/computational-cell-analytics/micro-sam/blob/master/micro_sam/util.py#L696 instead.

File: micro_sam/evaluation/inference.py [hunk 6]
```diff
@@ -392,118 +389,127 @@ def _save_segmentation(masks, prediction_path):
...
 def run_inference_with_iterative_prompting(
-    checkpoint_path: Union[str, os.PathLike],
-    model_type: str,
+    predictor: SamPredictor,
     image_paths: List[Union[str, os.PathLike]],
     gt_paths: List[Union[str, os.PathLike]],
-    prediction_root: Union[str, os.PathLike],
+    embedding_dir: Union[str, os.PathLike],
+    prediction_dir: Union[str, os.PathLike],
+    use_points: bool,
```
Comment (constantinpape) [inline micro_sam/evaluation/inference.py:474]: Having both `use_points` and `use_boxes` here are too many options. We should only have two choices:
- We start with a single positive point.
- We start with a bounding box.

So it's enough to have a single boolean for this, e.g. `start_with_box_prompt`.

File: micro_sam/evaluation/inference.py [hunk 7]
```diff
@@ -392,118 +389,127 @@ def _save_segmentation(masks, prediction_path):
...
+        embedding_dir: The directory where the image embeddings will be saved or are already saved.
+        prediction_dir: The directory where the predictions from SegmentAnything will be saved per iteration.
+        use_points: Whether to use the first prompt as 1 positive point
+        use_boxes: Whether to use the first prompt as bounding box
+        dilation: The dilation factor for the radius around the ground-truth obkect
+            around which points will not be sampled.
+        batch_size: The batch size used for batched predictions.
+        n_iterations: The number of iterations for iterative prompting.
+    """
+    if not (use_points or use_boxes):
```
Comment (constantinpape) [inline micro_sam/evaluation/inference.py:496]: We don't need this check if we simplify the choices here, see comment above.

File: micro_sam/evaluation/inference.py [hunk 8]
```diff
@@ -514,7 +520,10 @@ def run_inference_with_iterative_prompting(
...
         gt = relabel_sequential(gt)[0]
 
-        with torch.no_grad():
-            _run_inference_with_iterative_prompting_for_image(
-                model, image, gt, n_iterations, device, use_boxes, prediction_paths, batch_size,
-            )
+        embedding_path = os.path.join(embedding_dir, f"{os.path.splitext(image_name)[0]}.zarr")
+
+        _run_inference_with_iterative_prompting_for_image(
+            predictor, image, gt, use_points=use_points, use_boxes=use_boxes,
```
Comment (constantinpape) [inline micro_sam/evaluation/inference.py:526]: Change this to the simpler choice with just one boolean, see above.

File: micro_sam/evaluation/inference.py (hunk 5 above)
Comment (constantinpape) [reply micro_sam/evaluation/inference.py:436]: Note: this requires the gt to be of type `"int64"`, you can just pass it like this: `gt.astype("int64")` to make it work.

File: micro_sam/evaluation/inference.py [hunk 9]
```diff
@@ -392,118 +389,127 @@ def _save_segmentation(masks, prediction_path):
...
+    embedding_path,
+    n_iterations,
+    prediction_paths
+) -> None:
+    prompt_generator = IterativePromptGenerator()
 
-    image = torch.from_numpy(
-        image[None, None] if image.ndim == 2 else image[None]
+    gt_ids = np.unique(gt)[1:]
+    n_positives = 1 if not use_boxes else 0
```
Comment (constantinpape) [inline micro_sam/evaluation/inference.py:426]: You can simplify this a bit once we have the single boolean `start_with_box_prompt`:
```python
n_negatives = 0
if start_with_box_prompt:
  use_boxes, use_points = True, False
  n_positives = 0
  multimasking = False
else:
  use_boxes, use_points = False, True
  n_positives = 0
  multimasking = True
```

Comment (constantinpape) [reply micro_sam/evaluation/inference.py:426]: See also below for comment w.r.t. multi-masking.

File: micro_sam/evaluation/inference.py [hunk 10]
```diff
@@ -392,118 +389,127 @@ def _save_segmentation(masks, prediction_path):
...
-
-    multimasking = n_pos == 1
-    prompt_generator = IterativePromptGenerator()
-
-    n_samples = len(sampled_ids[0])
-    n_batches = int(np.ceil(float(n_samples) / batch_size))
+    # Use multi-masking only if we have a single positive point without box
+    multimasking = False
+    if not use_boxes and (n_positives == 1 and n_negatives == 0):
+        multimasking = True
```
Comment (constantinpape) [inline micro_sam/evaluation/inference.py:441]: You can make the mulitmasking decision above already, it's easier to read this if it's in a single if-else. And you need to always set it to False after the first loop, because we will then have more than one prompt and should not use multimasking anymore.

File: micro_sam/evaluation/inference.py [hunk 11]
```diff
@@ -392,118 +389,127 @@ def _save_segmentation(masks, prediction_path):
...
-                _inp["point_labels"] = updated_point_labels
-                _inp["mask_inputs"] = logits
+        if point_labels is not None:
+            point_labels = np.concatenate([point_labels, next_labels], axis=1)
+        else:
+            point_labels = next_labels
 
-        final_masks = torch.cat(final_masks, dim=1)
-        _save_segmentation(final_masks, prediction_paths[iteration])
+        _save_segmentation(masks, prediction_paths[iteration])
```
Comment (constantinpape) [inline micro_sam/evaluation/inference.py:465]: Set `multimasking = False` at the end of the loop. Otherwise it will stay true if it was set to true in the beginning. But that is wrong because we have more than one prompt after the first iteration.

**Review by constantinpape — COMMENTED** [review]
This looks good overall now! I left a few detailed comments, but these are mostly minor changes.

File: micro_sam/evaluation/inference.py (hunk 2 above)
Comment (anwai98) [reply micro_sam/evaluation/inference.py:398]: Ahha yes, that's correct. Fixed it, thanks!

File: micro_sam/evaluation/inference.py (hunk 5 above)
Comment (anwai98) [reply micro_sam/evaluation/inference.py:436]: Ahha good to know. I'll use this from here-on. Thanks!

File: micro_sam/evaluation/inference.py (hunk 11 above)
Comment (anwai98) [reply micro_sam/evaluation/inference.py:465]: Okay this is crucial. Thanks for pointing it out.

**Review by constantinpape — APPROVED** [review]
